stdlib/Random/src/ppmisc.jl

```diff
@@ -0,0 +1,167 @@
+export ppshuffle!, ppshuffle, pprandperm!, pprandperm
+
+## ppshuffle! & ppshuffle
+
+"""
+    _ppshuffle!(rng::TaskLocalRNG, B::AbstractArray{T}, A::Union{AbstractArray{T}, Base.OneTo{T}}, mask<:Union{UInt8, UInt16})
+
+Parallel Partitioned Shuffle
+1. partition input randomly
+2. shuffle partitions concurrently
+
+Arg `mask` determines number of partitions (mask + 1) to be used.
+"""
+function _ppshuffle!(r::TaskLocalRNG, B::AbstractArray{T}, A::Union{AbstractArray{T}, Base.OneTo{T}}, mask::Tu) where {T, Tu<:Union{UInt8, UInt16}}
+    # determine number of partitions
+    nparts = mask + 1
+    @assert ispow2(nparts) "invalid mask $(mask)"
+    @assert length(A) == length(B)
+
+    n = length(A)
+    s = Random.SamplerType{Tu}()
+
+    # an array to count partition hits by threads
+    # partitions map to rows (pid)
+    # threads map to cols (tid)
+    # we add an extra cache line for each thread
+    nrows = nparts < 8 ? 8 : nparts + 8
+    hits = zeros(Int, nrows, nthreads())
+
+    # save initial random state
+    r0 = copy(r)
+    # 1st pass
+    # assign input to partitions uniformly at random
+    # count cells hit by each thread in every partition
+    @threads :static for i in 1:n
+        local tid, pid = threadid(), rand(r, s) & mask + 1
+        @inbounds hits[pid, tid] += 1
+    end
+
+    # cumsum partition hits
+    # to mark boundaries of space reserved by each thread in every partition
+    # note that the 1st column will contain boundaries of entire partitions
+    prev = 0
+    for pid = 1:nparts, tid = 1:nthreads()
+        @inbounds prev = hits[pid, tid] += prev
+    end
+    # mark the end of the last partition
+    hits[nparts + 1, 1] = n
+
+    # recover random state
+    copy!(r, r0)
+    # 2nd pass
+    # scatter input accross partitions uniformly at random
+    # note that input distribution is identical as in the 1st pass
+    # since we recovered the initial random state
+    @threads :static for i in 1:n
+        local tid, pid = threadid(), rand(r, s) & mask + 1
+        @inbounds B[hits[pid, tid]] = A[i]
+        @inbounds hits[pid, tid] -= 1
+    end
+
+    # input is partitioned
+    # shuffle partitions in parallel
+    @threads :static for pid in 1:nparts
+        @inbounds local chunk = view(B, hits[pid, 1] + 1:hits[pid + 1, 1])
+        shuffle!(r, chunk)
+    end
+    B
+end
+
+"""
+    ppshuffle!([rng::TaskLocalRNG=default_rng(),] B::AbstractArray{T}, A::Union{AbstractArray{T}, Base.OneTo{T}}, mask<:Union{UInt8, UInt16})
+
+A multi-threaded implementation of [`shuffle!`](@ref).
+Construct in `B` a permuted copy of `A`.
+Optional arg `rng` specifies a random number generator (see [`TaskLocalRNG`](@ref)).
+
+# Examples
+```jldoctest
+julia> b = Vector{Int}(undef, 16);
+
+julia> ppshuffle!(b, 1:16);
+
+julia> isperm(b)
+true
+```
+"""
+function ppshuffle!(r::TaskLocalRNG, B::AbstractArray{T}, A::Union{AbstractArray{T}, Base.OneTo{T}}) where {T<:Integer}
+    nparts = max(2, (length(A) * sizeof(T)) >> 21)
```

Comment:
where does this come from?

Reply:
An experimental 'rule of thumb' setting an optimal partition size to 2MB and a minimal partition count to 2. Should be replaced by a more robust heuristic in future releases.